### PR TITLE
Fixed passing of delete method on button_to tag, creating wrong form csrf token

### LINKED ISSUE #23524: rails5 form authenticity_tokens don't work
This will cause problems for automation tools like mechanize.

Reproduction instructions: view source on any page that uses `form_for`.  Any edit page generated by scaffolding will do.  What you will see will be something like:

```
<!DOCTYPE html>
<html>
  <head>
    <title>Demo</title>
    <meta name="csrf-param" content="authenticity_token" />
<meta name="csrf-token" content="R/+8o7qCIsdpxYv0X/fxuZg8Tq5At6st3wmNM5vHgr7qdo29rBsTIE4mtxOFbGYTfcXTTI+Da88GirnPQUDkKQ==" />
    <meta name="action-cable-url" content="/cable" />

    <link rel="stylesheet" media="all" href="/assets/companies.self-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.css?body=1" data-turbolinks-track="true" />
<link rel="stylesheet" media="all" href="/assets/scaffolds.self-13e017ed781aee5641573fc553cae1177b6eb2ae02ee29ba9a3927350998c06d.css?body=1" data-turbolinks-track="true" />
<link rel="stylesheet" media="all" href="/assets/application.self-af04b226fd7202dfc532ce7aedb95a0128277937e90d3b3a3d35e1cce9e16886.css?body=1" data-turbolinks-track="true" />
    <script src="/assets/jquery.self-c64a74367bda6ef8b860f19e74df08927ca99d2be2ac934e9e92d5fd361e0da4.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/jquery_ujs.self-d602bdfe68ffc63b9f9cc512872aa3cfff046228a0a36e90dd476e8ef54c1b09.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/turbolinks.self-088c452541bc2cb3c0ff28ae33e5c6516d309f7e097363b1d7d062f9cbe5a947.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/action_cable.self-8fc7cfd5b796d0299c89a57dbf5b046051f42ce410f160fd7f1e677c8dc07def.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/action_cable/connection.self-5625494ff511c0f78cd0746ef7e14909f3d517ceb45f989511e4b216c1862593.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/action_cable/connection_monitor.self-ca8bfd8bf7c622d03bf8314cbde2cf77e5a7732c8082e2fc50828a14175f6b53.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/action_cable/subscriptions.self-156d38d6efaf3f6139b95c7fc9fcfc44a62d77f2c13108a2d58c33b0017eec10.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/action_cable/subscription.self-63bb8fcaf3e18e3bb4b849c831819921afe52d3e48e77f3c0d2d50fdf43a9b0c.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/action_cable/consumer.self-d0a6518de32602abb57c66be88b1fd0445527abfb6d8fad3a0be0c59eb6b7fd1.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/cable.self-877aef30ae1b040ab8a3aba4e3e309a11d7f2612f44dde450b5c157aa5f95c05.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/companies.self-877aef30ae1b040ab8a3aba4e3e309a11d7f2612f44dde450b5c157aa5f95c05.js?body=1" data-turbolinks-track="true"></script>
<script src="/assets/application.self-b89234cf2659d7fedea75bca0b8d231ad7dfc2f3f57fcbaf5f44ed9dc384137b.js?body=1" data-turbolinks-track="true"></script>
  </head>

  <body>
    <h1>Editing Company</h1>

<form class="edit_company" id="edit_company_2" action="/companies/2" accept-charset="UTF-8" method="post"><input name="utf8" type="hidden" value="&#x2713;" /><input type="hidden" name="_method" value="patch" /><input type="hidden" name="authenticity_token" value="1HxW3jw7tINiIbD8X25t15MukXOFV4g6sXD666f1oCt2BqGlZ1xGPHZ3w9NHuoISBFCynd63mETK8MHETZCTYw==" />

  <div class="field">
    <label for="company_name">Name</label>
    <input type="text" value="asdf" name="company[name]" id="company_name" />
  </div>

  <div class="actions">
    <input type="submit" name="commit" value="Update company" data-disable-with="Update company" />
  </div>
</form>

<a href="/companies/2">Show</a> |
<a href="/companies">Back</a>

  </body>
</html>
```

The important lines are:

```
<meta name="csrf-token" content="R/+8o7qCIsdpxYv0X/fxuZg8Tq5At6st3wmNM5vHgr7qdo29rBsTIE4mtxOFbGYTfcXTTI+Da88GirnPQUDkKQ==" />
```

and

```
<input type="hidden" name="authenticity_token" value="1HxW3jw7tINiIbD8X25t15MukXOFV4g6sXD666f1oCt2BqGlZ1xGPHZ3w9NHuoISBFCynd63mETK8MHETZCTYw==" />
```

Note that the values don't match.  They do with Rails 4.x releases.

### PR DESCRIPTION
**Pull request by vipulnsward**
@sikachu I took a stab at this, while looking around. :grin: 

r? @rafaelfranca 

Fixes #23524

### CODE REVIEW

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 1]
```diff
@@ -710,6 +714,46 @@ def test_rejects_token_for_incorrect_method
...
+    end
+
+    actual = @controller.send(:unmask_token, Base64.strict_decode64(form_token))
+    expected = @controller.send(:per_form_csrf_token, session, '/per_form_tokens/post_one', 'delete')
+    assert_equal expected, actual
+
+    # This is required because PATH_INFO isn't reset between requests.
+    @request.env['PATH_INFO'] = '/per_form_tokens/post_one'
+    assert_nothing_raised do
+      delete :post_one, params: {custom_authenticity_token: form_token}
```
Comment (kaspth) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Please add spaces between the braces everywhere: `params: { custom_authenticity_token: form_token }`.

Comment (vipulnsward) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Ha! I need to fix my editor for this.

Comment (vipulnsward) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Actually, I just copied this from above. This is used at many places this way.

Comment (kaspth) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Really? Then let's go with the file's precedent. Ugh.

File: actionview/lib/action_view/helpers/url_helper.rb [hunk 2]
```diff
@@ -311,8 +311,9 @@ def button_to(name = nil, options = nil, html_options = nil, &block)
         form_options[:action] = url
         form_options[:'data-remote'] = true if remote
 
-        request_token_tag = if form_method == 'post'
-          token_tag(nil, form_options: form_options)
+        token_tag_method = method == 'delete' ? 'delete' : form_method
+        request_token_tag = if (form_method == 'post' || method == 'delete')
+          token_tag(nil, form_options: form_options.merge(method: token_tag_method))
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/url_helper.rb]: We should just assign the method to `form_options` on line 310.

File: actionpack/test/controller/request_forgery_protection_test.rb (hunk 1 above)
Comment (vipulnsward) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Yup- http://take.ms/63V0s

File: actionview/lib/action_view/helpers/url_helper.rb (hunk 2 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/url_helper.rb]: Could even change 307 to:

```ruby
form_method  = method == 'get' || method == 'delete' ? method : 'post'
```

Comment (vipulnsward) [inline actionview/lib/action_view/helpers/url_helper.rb]: No, we still need that to be `post`

Comment (kaspth) [inline actionview/lib/action_view/helpers/url_helper.rb]: Yeah, just realized. Let's move `token_tag_method` into the `if` though.

Comment (kaspth) [inline actionview/lib/action_view/helpers/url_helper.rb]: Actually, those 3 lines are heavily condensed for no reason. Prefer some incidental duplication with:

```ruby
request_token_tag = \
  if form_method == 'post'
    token_tag(nil, form_options: form_options)
  elsif method == 'delete'
    token_tag(nil, form_options: form_options.merge(method: 'delete'))
  else
    ''
  end
```

**Comment by kaspth** [comment]
Doesn't `button_to` also support PUT and PATCH?

File: actionview/lib/action_view/helpers/url_helper.rb [hunk 3]
```diff
@@ -311,8 +311,9 @@ def button_to(name = nil, options = nil, html_options = nil, &block)
         form_options[:action] = url
         form_options[:'data-remote'] = true if remote
 
-        request_token_tag = if form_method == 'post'
-          token_tag(nil, form_options: form_options)
```
Comment (rafaelfranca) [inline actionview/lib/action_view/helpers/url_helper.rb]: The only change we need is in this line to `token_tag(nil, form_options: { action: url, method: method })`

**Comment by rafaelfranca** [comment]
> Doesn't button_to also support PUT and PATCH?

Yes. I commented with the code that would make all methods work. We need tests for put and patch too.

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 4]
```diff
@@ -710,6 +714,46 @@ def test_rejects_token_for_incorrect_method
     end
   end
 
+  def test_rejects_token_for_incorrect_method_button_to
+    get :button_to, params: {form_method: 'delete'}
```
Comment (rafaelfranca) [inline actionpack/test/controller/request_forgery_protection_test.rb]: space after the `{` and before `}`

File: actionpack/test/controller/request_forgery_protection_test.rb (hunk 1 above)
Comment (rafaelfranca) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Even if the above method is using wrong style new code should use only the right style.

File: actionview/lib/action_view/helpers/url_helper.rb (hunk 3 above)
Comment (vipulnsward) [inline actionview/lib/action_view/helpers/url_helper.rb]: :+1:

**Comment by vipulnsward** [comment]
All done here.

**Comment by kaspth** [comment]
There's still a little ways to go. This code doesn't support PUT or PATCH verbs, we need those too plus tests for them.

Luckily, this is easy if we follow @rafaelfranca's suggestion. We don't have to compare `method == 'delete'` or any other verb, we only care if the `form_method` is POST. If so, just pass `method` straight on through to `token_tag`. That's what Rafael meant with:

> The only change we need is in this line to token_tag(nil, form_options: { action: url, method: method })

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 5]
```diff
@@ -710,6 +714,46 @@ def test_rejects_token_for_incorrect_method
...
+
+    form_token = nil
+    assert_select 'input[name=custom_authenticity_token]' do |elts|
+      form_token = elts.first['value']
+      assert_not_nil form_token
+    end
+
+    actual = @controller.send(:unmask_token, Base64.strict_decode64(form_token))
+    expected = @controller.send(:per_form_csrf_token, session, '/per_form_tokens/post_one', 'delete')
+    assert_equal expected, actual
```
Comment (kaspth) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Hard to track the purpose of these 8 lines comparing the generated token to the one on the server. Could do well with a custom assertion. E.g. `assert_matches_session_token form_token`. With that we'd need a helper to extract the `form_token` ala:

```ruby
def form_csrf_token
  assert_select 'input[name="custom_authenticity_token]" do |input|
    return input.first['value']
  end
end
```

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 6]
```diff
@@ -710,6 +714,46 @@ def test_rejects_token_for_incorrect_method
     end
   end
 
+  def test_rejects_token_for_incorrect_method_button_to
+    get :button_to, params: { form_method: 'delete' }
+
+    form_token = nil
+    assert_select 'input[name=custom_authenticity_token]' do |elts|
+      form_token = elts.first['value']
+      assert_not_nil form_token
```
Comment (kaspth) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Would move this into `assert_matches_session_token`

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 7]
```diff
@@ -710,6 +714,46 @@ def test_rejects_token_for_incorrect_method
     end
   end
 
+  def test_rejects_token_for_incorrect_method_button_to
+    get :button_to, params: { form_method: 'delete' }
+
+    form_token = nil
+    assert_select 'input[name=custom_authenticity_token]' do |elts|
```
Comment (kaspth) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Nokogiri doesn't like selecting on params without quotes, switch to: `'input[name="custom_authenticity_token"]'`

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 8]
```diff
@@ -710,6 +714,46 @@ def test_rejects_token_for_incorrect_method
...
+      assert_not_nil form_token
+    end
+
+    actual = @controller.send(:unmask_token, Base64.strict_decode64(form_token))
+    expected = @controller.send(:per_form_csrf_token, session, '/per_form_tokens/post_one', 'delete')
+    assert_equal expected, actual
+
+    # This is required because PATH_INFO isn't reset between requests.
+    @request.env['PATH_INFO'] = '/per_form_tokens/post_one'
+    assert_nothing_raised do
```
Comment (kaspth) [inline actionpack/test/controller/request_forgery_protection_test.rb]: We can't make other assertions than `assert_nothing_raised`?

File: actionview/lib/action_view/helpers/url_helper.rb [hunk 9]
```diff
@@ -311,8 +311,8 @@ def button_to(name = nil, options = nil, html_options = nil, &block)
...
+        request_token_tag = if (form_method == 'post' || method == 'delete')
```
Comment (rafaelfranca) [inline actionview/lib/action_view/helpers/url_helper.rb]: We don't need to add the `||` here.

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 10]
```diff
@@ -710,6 +714,46 @@ def test_rejects_token_for_incorrect_method
...
+
+    # This is required because PATH_INFO isn't reset between requests.
+    @request.env['PATH_INFO'] = '/per_form_tokens/post_one'
+    assert_raises(ActionController::InvalidAuthenticityToken) do
+      patch :post_one, params: { custom_authenticity_token: form_token }
+    end
+  end
+
+  def test_accepts_proper_token_for_delete_method_button_to
+    get :button_to, params: { form_method: 'delete' }
```
Comment (rafaelfranca) [inline actionpack/test/controller/request_forgery_protection_test.rb]: We need to test with put and patch too.

**Comment by vipulnsward** [comment]
Refactored the tests to dry them up, into meaningful chunks.

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 11]
```diff
@@ -794,4 +791,20 @@ def test_method_is_case_insensitive
...
+      return form_csrf_token
+    end
+  end
+
+  def assert_matches_session_token_on_server(form_token, method = 'post')
+    actual = @controller.send(:unmask_token, Base64.strict_decode64(form_token))
+    expected = @controller.send(:per_form_csrf_token, session, '/per_form_tokens/post_one', method)
+    assert_equal expected, actual
+  end
+
```
Comment (eileencodes) [inline actionpack/test/controller/request_forgery_protection_test.rb]: Nitpick: can you remove this extra space here, and you have 2 spaces above where "assert_presence_and_fetch_form_csrf_token" starts. Can you remove those too?

File: actionpack/test/controller/request_forgery_protection_test.rb [hunk 12]
```diff
@@ -794,4 +791,19 @@ def test_method_is_case_insensitive
...
+        assert_not_nil form_csrf_token
+        return form_csrf_token
+      end
+    end
+
+    def assert_matches_session_token_on_server(form_token, method = 'post')
+      actual = @controller.send(:unmask_token, Base64.strict_decode64(form_token))
+      expected = @controller.send(:per_form_csrf_token, session, '/per_form_tokens/post_one', method)
+      assert_equal expected, actual
+    end
```
Comment (vipulnsward) [inline actionpack/test/controller/request_forgery_protection_test.rb]: @eileencodes done, marked as private as well.

**Pull request merged**